ktor-utils/common/src/io/ktor/util/pipeline/PipelineContext.kt

```diff
@@ -84,28 +84,44 @@ private class SuspendFunctionGun<TSubject : Any, TContext : Any>(
 
         private fun peekContinuation(): Continuation<*>? {
             if (lastPeekedIndex < 0) return null
-            val rootContinuation = rootContinuation
 
-            @Suppress("UNCHECKED_CAST")
-            when (rootContinuation) {
+            when (val rootContinuation = rootContinuation) {
                 null -> return null
                 is Continuation<*> -> {
                     --lastPeekedIndex
                     return rootContinuation
                 }
                 is ArrayList<*> -> {
-                    if (rootContinuation.isEmpty()) return null
-                    return rootContinuation[lastPeekedIndex--] as Continuation<*>
+                    if (rootContinuation.isEmpty()) return StackWalkingFailedFrame
```

Comment:
Isn't empty list a valid situation?

Reply:
Right, it is.

Reply:
Well, the answer is both yes and no. If the list is empty, then `lastPeekedIndex` should be `-1` and `null` should be already returned at the beginning of the function. So if we are here and see that the list is empty but lastPeekedIndex wasn't `-1`, then it means that element(s) has disappeared concurrently. 

I think we have two options here: we may return `null` and say that it's ok in spite of the accident, or we may report it as a problem as it is done in this PR.